crypto/x509/x509_vfy.c

```diff
@@ -1806,9 +1804,20 @@ int ossl_x509_check_cert_time(X509_STORE_CTX *ctx, X509 *x, int depth)
  */
 static int internal_verify(X509_STORE_CTX *ctx)
 {
-    int n = sk_X509_num(ctx->chain) - 1;
-    X509 *xi = sk_X509_value(ctx->chain, n);
-    X509 *xs = xi;
+    int n;
+    X509 *xi;
+    X509 *xs;
+
+    /* For RPK: just do the verify callback */
+    if (ctx->rpk != NULL) {
+        /* TODO: should this be (0, ctx)? since we can't validate an RPK except via DANE? */
+        if (!ctx->verify_cb(1, ctx))
```

Comment:
Here, I am not sure whether to call the "traditional" verification callback, with users who enable RPK expected to have callbacks that can deal with absence of certificates, or to add a new callback for RPK. Thoughts?

Reply:
I suppose it would depend on exactly what the defined behavior for the callback is.

By the way, DANE technically isn't the only way to validate an RPK: a client could be explicitly configured to trust that public key, for instance.

Reply:
> By the way, DANE technically isn't the only way to validate an RPK: a client could be explicitly configured to trust that public key, for instance.

You've missed the discussion upthread, you're confusing verification via data in the form of TLSA records as a mechanism for specifying trust anchors with using DNS(SEC) as a source of said records. The direction proposed here is to use TLSA records regardless of whether the ultimately come from DNS, or are just a means of matching a particular key.  In that sense, TLSA records are *universal*, they support either local policy based trust anchors, or TLSA records from some external source, typically found via DNSSEC.

Reply:
DANE it OpenSSL is source-agnostic. Postfix, for example, uses the DANE API in OpenSSL for both actual DNSSEC-based DANE, and for its "fingerprint" security level, which matches peer certificates against local pins of either the certificate or the key.

Reply:
> I wouldn't say using TLSA outside the context of DNSSEC is DANE.
> 
> DANE is using DNSSEC to validate certificates. It isn't even specific to TLSA records: SSHFP is also DANE, for instance.

Please desist. This is a forum for reviewing the pull request, not nitpicking terminology. Support for DANE in OpenSSL is support for using data in the form of TLSA records to verify the presented peer credentials (currently X.509 certificates, and when this WIP PR is done also raw public keys).  The TLSA records can be synthesised from various other input forms (DER SPKI public keys, SHA2-256 digests of same DER form, EE or issuer certificate date when not RPK, ...). Further comments that distract from the PR will be deleted.

Reply:
> Here, I am not sure whether to call the "traditional" verification callback, with users who enable RPK expected to have callbacks that can deal with absence of certificates, or to add a new callback for RPK. Thoughts?

At this stage; I'm using the existing callback. Existing applications should not be impacted as they need to explicitly enable certificate type configuration, which presumably would mean authors need to update the verification callback. The verification callback is not specific to a certificate (i.e. no explicit X509 argument), so it could be used for either RPK or certificates without an interface change.

Again, WIP, but I'm making progress.

Reply:
Yes, for now now issues. It is common for application callbacks to inspect the certificate, for logging, ... but yes enabling RPK could be handled with a bit of care in the same callback in applications that support RPK. Perhaps that's sufficient.